File: src/shared/api/im.ts
import { invokeLogged, isTauriRuntime } from "./runtime";
import { formatLocalDateTime } from "../id";
import { browserMock, cloneImSettings, getImProvider } from "../mock/browser";
import {
  FeishuCredentialStatus,
  FeishuGatewayStatus,
  ImGatewayStatus,
  ImIntegrationSettings,
  ImLoginStatus,
  ImProviderCredentialStatus,
  ImProviderId,
} from "../types";

/** 读取即时通讯设置；浏览器开发态返回内存 mock。 */
export async function loadImSettings(): Promise<ImIntegrationSettings> {
  if (!isTauriRuntime()) {
    return cloneImSettings(browserMock.imSettings);
  }

  return invokeLogged<ImIntegrationSettings>("load_im_settings");
}

/** 保存即时通讯设置；敏感凭证由独立 keyring 命令处理。 */
export async function saveImSettings(settings: ImIntegrationSettings): Promise<ImIntegrationSettings> {
  if (!isTauriRuntime()) {
    browserMock.imSettings = cloneImSettings(settings);
    for (const provider of settings.providers) {
      const current = browserMock.imGatewayByProvider[provider.providerId] ?? browserMock.imGatewayByProvider.feishu;
      const identityConfigured =
        provider.config.type === "feishu"
          ? Boolean(provider.config.appId.trim())
          : provider.config.type === "qq"
            ? Boolean(provider.config.appId.trim())
            : Boolean(provider.config.accountId.trim());
      browserMock.imGatewayByProvider[provider.providerId] = {
        ...current,
        providerId: provider.providerId,
        domain: provider.config.type === "feishu" ? provider.config.domain : provider.providerId,
        appIdConfigured: identityConfigured,
      };
    }
    browserMock.feishuGatewayStatus = browserMock.imGatewayByProvider.feishu;

    return loadImSettings();
  }

  return invokeLogged<ImIntegrationSettings>("save_im_settings", { payload: { settings } });
}

/** 保存 IM provider secret；桌面端写入系统安全存储，浏览器态只返回不可用说明。 */
export async function saveImProviderSecret(providerId: ImProviderId, secret: string): Promise<ImProviderCredentialStatus> {
  if (!isTauriRuntime()) {
    throw new Error("浏览器开发态不能保存 IM provider secret，请在 Tauri 桌面端配置。");
  }

  return invokeLogged<ImProviderCredentialStatus>("save_im_provider_secret", { payload: { providerId, secret } });
}

/** 读取 IM provider secret 是否已配置；不会返回明文 secret。 */
export async function loadImProviderCredentialStatus(providerId: ImProviderId): Promise<ImProviderCredentialStatus> {
  if (!isTauriRuntime()) {
    const provider = getImProvider(browserMock.imSettings, providerId);

    return {
      providerId,
      keyReference: provider.config.secretKeyReference,
      configured: false,
      message: "浏览器开发态未连接系统安全存储。",
    };
  }

  return invokeLogged<ImProviderCredentialStatus>("load_im_provider_credential_status", { payload: { providerId } });
}

/** 启动 IM provider 长连接网关；浏览器态只返回不可用状态。 */
export async function startImGateway(providerId: ImProviderId): Promise<ImGatewayStatus> {
  if (!isTauriRuntime()) {
    browserMock.imGatewayByProvider[providerId] = {
      ...browserMock.imGatewayByProvider[providerId],
      providerId,
      running: false,
      connected: false,
      lastError: "浏览器开发态不能启动 IM 长连接网关。",
    };
    browserMock.feishuGatewayStatus = browserMock.imGatewayByProvider.feishu;

    return browserMock.imGatewayByProvider[providerId];
  }

  return invokeLogged<ImGatewayStatus>("start_im_gateway", { payload: { providerId } });
}

/** 停止 IM provider 长连接网关；不会清空设置和凭证。 */
export async function stopImGateway(providerId: ImProviderId): Promise<ImGatewayStatus> {
  if (!isTauriRuntime()) {
    browserMock.imGatewayByProvider[providerId] = {
      ...browserMock.imGatewayByProvider[providerId],
      providerId,
      running: false,
      connected: false,
      lastStoppedAt: formatLocalDateTime(),
    };
    browserMock.feishuGatewayStatus = browserMock.imGatewayByProvider.feishu;

    return browserMock.imGatewayByProvider[providerId];
  }

  return invokeLogged<ImGatewayStatus>("stop_im_gateway", { payload: { providerId } });
}

/** 读取 IM provider 长连接网关运行态。 */
export async function loadImGatewayStatus(providerId: ImProviderId): Promise<ImGatewayStatus> {
  if (!isTauriRuntime()) {
    return { ...browserMock.imGatewayByProvider[providerId], providerId };
  }

  return invokeLogged<ImGatewayStatus>("load_im_gateway_status", { payload: { providerId } });
}

/** 保存飞书 appSecret；兼容旧调用，内部走通用 IM provider secret 命令。 */
export async function saveFeishuAppSecret(appSecret: string): Promise<FeishuCredentialStatus> {
  return saveImProviderSecret("feishu", appSecret);
}

/** 读取飞书 appSecret 是否已配置；兼容旧调用。 */
export async function loadFeishuCredentialStatus(): Promise<FeishuCredentialStatus> {
  return loadImProviderCredentialStatus("feishu");
}

/** 启动飞书长连接网关；兼容旧调用。 */
export async function startFeishuGateway(): Promise<FeishuGatewayStatus> {
  return startImGateway("feishu");
}

/** 停止飞书长连接网关；兼容旧调用。 */
export async function stopFeishuGateway(): Promise<FeishuGatewayStatus> {
  return stopImGateway("feishu");
}

/** 读取飞书长连接网关运行态；兼容旧调用。 */
export async function loadFeishuGatewayStatus(): Promise<FeishuGatewayStatus> {
  return loadImGatewayStatus("feishu");
}

function emptyLoginStatus(providerId: ImProviderId, message: string): ImLoginStatus {
  return {
    providerId,
    status: "idle",
    message,
  };
}

/** 启动 IM 扫码登录；浏览器开发态不连接真实接口。 */
export async function startImLogin(providerId: ImProviderId): Promise<ImLoginStatus> {
  if (!isTauriRuntime()) {
    return emptyLoginStatus(providerId, "浏览器开发态不能扫码登录，请在 Tauri 桌面端操作。");
  }

  return invokeLogged<ImLoginStatus>("start_im_login", { payload: { providerId } });
}

/** 读取 IM 扫码登录状态。 */
export async function loadImLoginStatus(providerId: ImProviderId): Promise<ImLoginStatus> {
  if (!isTauriRuntime()) {
    return emptyLoginStatus(providerId, "浏览器开发态未连接扫码登录。");
  }

  return invokeLogged<ImLoginStatus>("load_im_login_status", { payload: { providerId } });
}

/** 取消进行中的 IM 扫码登录。 */
export async function cancelImLogin(providerId: ImProviderId): Promise<ImLoginStatus> {
  if (!isTauriRuntime()) {
    return emptyLoginStatus(providerId, "浏览器开发态没有进行中的扫码登录。");
  }

  return invokeLogged<ImLoginStatus>("cancel_im_login", { payload: { providerId } });
}

File: src-tauri/src/commands/im.rs
use super::common::*;

#[tauri::command]
pub async fn load_im_settings(app: AppHandle) -> Result<ImIntegrationSettings, String> {
    run_blocking("读取即时通讯设置", move || {
        storage::load_im_settings(&app)
    })
    .await
}
/** 保存即时通讯集成设置；飞书 appSecret 必须走独立 keyring 命令。 */
#[tauri::command]
pub async fn save_im_settings(
    app: AppHandle,
    payload: SaveImSettingsPayload,
) -> Result<ImIntegrationSettings, String> {
    let settings_app = app.clone();
    let started_at = Instant::now();
    let result = run_blocking("保存即时通讯设置", move || {
        storage::save_im_settings(&settings_app, &payload.settings)
    })
    .await;

    match &result {
        Ok(settings) => logging::write_app_event_best_effort(
            &app,
            AppEventBuilder::new(
                AppLogLevel::Info,
                AppLogCategory::Im,
                "save_im_settings",
                "completed",
                "已保存即时通讯设置。",
            )
            .duration(started_at.elapsed())
            .metadata(json!({
                "providerCount": settings.providers.len(),
                "enabledProviderCount": settings.providers.iter().filter(|provider| provider.enabled).count(),
                "providers": settings.providers.iter().map(|provider| json!({
                    "providerId": provider.provider_id,
                    "enabled": provider.enabled,
                    "knowledgeBaseCount": provider.default_knowledge_base_ids.len(),
                    "allowedUserCount": provider.allowed_user_open_ids.len(),
                    "allowedChatCount": provider.allowed_chat_ids.len(),
                    "requireMention": provider.require_mention,
                })).collect::<Vec<_>>(),
            })),
        ),
        Err(error) => logging::write_app_event_best_effort(
            &app,
            AppEventBuilder::new(
                AppLogLevel::Error,
                AppLogCategory::Im,
                "save_im_settings",
                "failed",
                error,
            )
            .duration(started_at.elapsed()),
        ),
    }

    result
}

/** 保存 IM provider secret 到系统安全存储；命令日志不包含 secret 明文。 */
#[tauri::command]
pub async fn save_im_provider_secret(
    app: AppHandle,
    payload: SaveImProviderSecretPayload,
) -> Result<ImProviderCredentialStatus, String> {
    let provider_id = payload.provider_id.trim().to_ascii_lowercase();
    let started_at = Instant::now();
    let result_provider_id = provider_id.clone();
    let result = run_blocking("保存 IM provider secret", move || {
        storage::save_im_provider_secret(&provider_id, &payload.secret)
    })
    .await;

    match &result {
        Ok(_) => logging::write_app_event_best_effort(
            &app,
            AppEventBuilder::new(
                AppLogLevel::Info,
                AppLogCategory::Im,
                "save_im_provider_secret",
                "completed",
                "IM provider secret 已保存。",
            )
            .duration(started_at.elapsed())
            .metadata(json!({ "providerId": result_provider_id })),
        ),
        Err(error) => logging::write_app_event_best_effort(
            &app,
            AppEventBuilder::new(
                AppLogLevel::Error,
                AppLogCategory::Im,
                "save_im_provider_secret",
                "failed",
                error,
            )
            .duration(started_at.elapsed())
            .metadata(json!({ "providerId": result_provider_id })),
        ),
    }

    result
}

/** 读取 IM provider secret 是否已配置；不返回明文 secret。 */
#[tauri::command]
pub async fn load_im_provider_credential_status(
    payload: ImProviderPayload,
) -> Result<ImProviderCredentialStatus, String> {
    let provider_id = payload.provider_id.trim().to_ascii_lowercase();

    run_blocking("读取 IM provider 凭证状态", move || {
        storage::load_im_provider_credential_status(&provider_id)
    })
    .await
}

/** 启动 IM provider 网关，消息处理进入后台任务。 */
#[tauri::command]
pub async fn start_im_gateway(
    app: AppHandle,
    payload: ImProviderPayload,
) -> Result<ImGatewayStatus, String> {
    let provider_id = payload.provider_id.trim().to_ascii_lowercase();

    crate::im::start_gateway(app, &provider_id).await
}

/** 停止 IM provider 网关，不清空任何配置或凭证。 */
#[tauri::command]
pub async fn stop_im_gateway(
    app: AppHandle,
    payload: ImProviderPayload,
) -> Result<ImGatewayStatus, String> {
    let provider_id = payload.provider_id.trim().to_ascii_lowercase();

    run_blocking("停止 IM provider 网关", move || {
        crate::im::stop_gateway(&app, &provider_id)
    })
    .await
}

/** 读取 IM provider 网关运行态。 */
#[tauri::command]
pub async fn load_im_gateway_status(
    app: AppHandle,
    payload: ImProviderPayload,
) -> Result<ImGatewayStatus, String> {
    let provider_id = payload.provider_id.trim().to_ascii_lowercase();

    run_blocking("读取 IM provider 网关状态", move || {
        crate::im::load_gateway_status(&app, &provider_id)
    })
    .await
}

/** 保存飞书 appSecret 到系统安全存储；兼容旧命令，内部转发到通用 provider 命令。 */
#[tauri::command]
pub async fn save_feishu_app_secret(
    app: AppHandle,
    payload: SaveFeishuSecretPayload,
) -> Result<FeishuCredentialStatus, String> {
    let started_at = Instant::now();
    let result = run_blocking("保存飞书 appSecret", move || {
        storage::save_im_provider_secret(IM_PROVIDER_FEISHU, &payload.app_secret)
    })
    .await;

    match &result {
        Ok(_) => logging::write_app_event_best_effort(
            &app,
            AppEventBuilder::new(
                AppLogLevel::Info,
                AppLogCategory::Im,
                "save_feishu_app_secret",
                "completed",
                "飞书 appSecret 已保存。",
            )
            .duration(started_at.elapsed()),
        ),
        Err(error) => logging::write_app_event_best_effort(
            &app,
            AppEventBuilder::new(
                AppLogLevel::Error,
                AppLogCategory::Im,
                "save_feishu_app_secret",
                "failed",
                error,
            )
            .duration(started_at.elapsed()),
        ),
    }

    result
}

/** 读取飞书 appSecret 是否已配置；不返回明文 secret。 */
#[tauri::command]
pub async fn load_feishu_credential_status() -> Result<FeishuCredentialStatus, String> {
    run_blocking("读取飞书凭证状态", || {
        storage::load_im_provider_credential_status(IM_PROVIDER_FEISHU)
    })
    .await
}

/** 启动飞书长连接网关；兼容旧命令，内部转发到通用 provider 路由。 */
#[tauri::command]
pub async fn start_feishu_gateway(app: AppHandle) -> Result<FeishuGatewayStatus, String> {
    crate::im::start_gateway(app, IM_PROVIDER_FEISHU).await
}

/** 停止飞书长连接网关；兼容旧命令，不清空任何配置或凭证。 */
#[tauri::command]
pub async fn stop_feishu_gateway(app: AppHandle) -> Result<FeishuGatewayStatus, String> {
    run_blocking("停止飞书长连接网关", move || {
        crate::im::stop_gateway(&app, IM_PROVIDER_FEISHU)
    })
    .await
}

/** 读取飞书长连接网关运行态；兼容旧命令。 */
#[tauri::command]
pub async fn load_feishu_gateway_status(app: AppHandle) -> Result<FeishuGatewayStatus, String> {
    run_blocking("读取飞书网关状态", move || {
        crate::im::load_gateway_status(&app, IM_PROVIDER_FEISHU)
    })
    .await
}
